repack-in-place optimizations

JamesPiechota opened the pull request.

JamesPiechota committed:
refactor: pre-compute entropy offsets rather than iterate

This will allow other modules to use the same set of offsets

JamesPiechota committed:
fix: batch repack_in_place based on entropy offsets

This allows us to have one read and one write per chunk repacked.

1. generate offsets covered by one 256 MiB entropy
2. read those chunks from disk and repack them, cache the results
3. generate the 256 MiB entropy, cache the results
4. once entropy is generated and chunks read, encipher all chunks an
   write them back to disk

JamesPiechota committed:
fix: repack in place multiple entropy footprints at a time

`repack_batch_size` controls how many 256 MiB entropy footprints
are repacked at once. The larger the batch size the more chunks
can be read at once, reducing the read io bottleneck

JamesPiechota committed:
test: add some basic ar_repack unit tests

JamesPiechota committed:
fix: add a repack state machine

this does two things
1. makes it much easier to keep track of what's happening when
2. enable breaking down each phase into smaller parts. e.g. instead
   of having to read all chunks at once and then do the next steps
   we can read in batches and allow other steps to happen concurrently.
   This greatly reduces the memory load.

JamesPiechota committed:
fix: break out repack-in-place IO to its own process

This helps maximize disk utilization by maintining a single
read/write queue that doesn't get interrupted by other
repack operations

JamesPiechota committed:
fix: add some metrics to better track repack process

JamesPiechota committed:
feat: add 'data-doctor inspect bitmap'

Dumps a bitmap of a storage module showing the packing format
of each chunk

JamesPiechota committed:
fix: repack-in-place cursor logic

Avoid reprocessing already processed chunks, also invalidate
any chunks we detect with mismatched sync records

JamesPiechota committed:
docs: document the repack fsm

JamesPiechota committed:
refactor: rename a few concepts in ar_repack

- footprint:  represents both the set of offsets covered when we
              generate entropy and now also represents one iteration
              of the repack process. A repack footprint is N entropy
              footprints where N  is the repack batch size.
- batch:      represents the number of consecutive chunks read at
              a time. Set with read_batch_size
- read range: the range of chunks read

JamesPiechota committed:
refactor: rename chunk_info to repack_chunk

ldmberman commented on apps/arweave/src/ar_sync_record.erl:
```diff
@@ -131,8 +132,10 @@ delete(End, Start, ID, StoreID) ->
+delete_async(Event, End, Start, ID, StoreID, ReportErrorsTo) ->
 	GenServerID = name(StoreID),
-	gen_server:cast(GenServerID, {delete_async, Event, End, Start, ID}).
+	gen_server:cast(GenServerID, {delete_async, Event, End, Start, ID, ReportErrorsTo}).
```
Did you add ReportErrorsTo for debugging purposes?

JamesPiechota replied:
I can probably remove this. In an earlier iteration I'd used delete_async to try to optimize the repacking. And then added REportErrorsTo as a way to be notified if there was a problem so that ar_repack could then invalidated that chunk (rather than leave it in an intermediate state)

However I backed that change out cause it didn't make a difference in performance but dded a lot of complexity. I just did a quick scan and it turns out we don't use delete_async anywhere in the codebase. So maybe I'll just remove the function altogether?

Commit:
refactor: code cleanup

JamesPiechota committed:
feat: expose repack_cache_size_mb as a configuration option

There are 4 values which control repack-in-place performance
and resource use:

- exposed: repack_batch_size: number of consecutive chunks to read at a time
- exposed: repack_cache_size_mb: maximum size of the repack chunk cache
- internal: write_batch_size: number of chunks to write at a time.
                              hardcoded to 1024
- internal: num_entropy_offsets: how much entropy to use at once

By default when we generate entropy we generate 1024 chunks worth.
But this creates some memory pressure since if we want a repack_batch_size
of 100 (for example) that means we need to cache 100 different entropy
footprints for a total size of 102,400 chunks, or ~25GiB. To address this
we can set the num_entropy_offsets less than 1024. This means we will
generate some entropy and then throw it away. For example if
num_entropy_offsets is 128, we only use 128 chunks of entropy each time
instead of the full 1024, and so will end up calculating the same entropy
8 times. Then node will set num_entropy_offsets automatically in order
to satisfy the repack_batch_size and repack_cache_size_mb.

This can be a useful config option to balance RAM vs. Disk IO vs. CPU

Commit:
test: fix ar_entropy_storage tests

JamesPiechota committed:
refactor: remove ar_sync_record:delete_async

was not used anywhere

JamesPiechota committed:
refactor: expand use of #chunk_metadata and #chunk_proof (#765)

This is an incremental step towards replacing some of our long,
opaque tuple arguments with records that can be used across
modules

JamesPiechota committed:
test: number of fixes for flaky tests

JamesPiechota committed:
fix: generate entropy in sequence instead of all at once

This reduces the maximum memory required during repack in place.

JamesPiechota committed:
fix: when collecting peers adjust disk space delay

if disk space is not sufficient, we delay collecting peers for
a bit. Reduce teh delay if we're waiting to join the network (this
also better handles some testing flakiness)

Commit:
refactor: code cleanup + handle entropy generation timeout

JamesPiechota commented on apps/arweave/src/ar_repack.erl:
```diff
@@ -1,384 +1,967 @@
+			update_chunk_state(RepackChunk2, State2);
+		already_repacked ->
+			%% Remove the chunk and entropy to free up memory.
+			RepackChunk2 = RepackChunk#repack_chunk{ chunk = <<>> },
```
Ah yeah. It's a bad comment. If we're in the `already_repacked` state it means that the `entropy` hasn't been set yet. Once the `entropy` is set we'll move to the `ignore` state and the RepackChunk will be removed from the map entirely. I'll update the comment.

JamesPiechota replied:
https://github.com/ArweaveTeam/arweave/pull/712/commits/7c1b6cdea361204a31d34042dc9f03f59d19e42b
